### PR TITLE
Cap the number of glyphs per WebRender text run to avoid overflowing WR's limit.

### LINKED ISSUE #17230: Intermittent crash in /dom/nodes/Document-contentType/contentType/contenttype_html.html, /dom/nodes/Document-contentType/contentType/contenttype_txt.html, /dom/nodes/Document-contentType/contentType/contenttype_xml.html ("request.current_used_block_num() <= MAX_VERTEX_TEXTURE_WIDTH")
```
  ▶ CRASH [expected OK] /dom/nodes/Document-contentType/contentType/contenttype_html.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ Can't allocate > MAX_VERTEX_TEXTURE_WIDTH per resource! (thread RenderBackend, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/caec344/webrender/src/gpu_cache.rs:240)
  │ 2017-06-08 06:23:41.228 servo[34660:1658180238] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x1080f46ee - backtrace::backtrace::trace::h15d769766aa71ae0
  │    1:        0x1080f487c - backtrace::capture::Backtrace::new::hde1afc496271e12d
  │    2:        0x106caa666 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::he3b0f9713346a53e
  │    3:        0x1089f05f6 - std::panicking::rust_panic_with_hook::h8b9b25777425677b
  │    4:        0x108194634 - std::panicking::begin_panic::h0180222598f02c7d
  │    5:        0x10821992a - _$LT$webrender..gpu_cache..GpuDataRequest$LT$$u27$a$GT$$u20$as$u20$core..ops..Drop$GT$::drop::h39e0b81640f596f9
  │    6:        0x10821cf20 - _$LT$webrender..prim_store..TextRunPrimitiveCpu$u20$as$u20$webrender..gpu_cache..ToGpuBlocks$GT$::write_gpu_blocks::hc5b7908553c4ba22
  │    7:        0x108216c87 - webrender::frame_builder::LayerRectCalculationAndCullingPass::handle_primitive_run::h3433b24e91763df0
  │    8:        0x10820cabf - webrender::frame_builder::FrameBuilder::build::h82fe04f16c0fd444
  │    9:        0x10822d4b1 - webrender::render_backend::RenderBackend::render::hc7ba8677c3d5fe77
  │   10:        0x108222348 - webrender::render_backend::RenderBackend::run::h7d249d7284265167
  │   11:        0x108192c55 - std::sys_common::backtrace::__rust_begin_short_backtrace::h11923530aed18a36
  │   12:        0x10819588d - std::panicking::try::do_call::h7dc0a5e886a59d78
  │   13:        0x1089f172a - __rust_maybe_catch_panic
  │   14:        0x1081b37a6 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::h9765d4eb192e382f
  │   15:        0x1089ed4c5 - std::sys::imp::thread::Thread::new::thread_start::h993b7e5cb7871f66
  │   16:     0x7fff95aed059 - _pthread_body
  │   17:     0x7fff95aecfd6 - _pthread_start
  │ ERROR:servo: Can't allocate > MAX_VERTEX_TEXTURE_WIDTH per resource!
  │ called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) }) (thread main, at src/libcore/result.rs:859)
  │ called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) }) (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(2) }, at src/libcore/result.rs:859)
  │ stack backtrace:
  │    0:        0x1080f46ee - backtrace::backtrace::trace::h15d769766aa71ae0
  │    1:        0x1080f487c - backtrace::capture::Backtrace::new::hde1afc496271e12d
  │    2:        0x106caa666 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::he3b0f9713346a53e
  │    3:        0x1089f05f6 - std::panicking::rust_panic_with_hook::h8b9b25777425677b
  │    4:        0x1089f0454 - std::panicking::begin_panic::h69fe06a9d67b6165
  │    5:        0x1089f03c2 - std::panicking::begin_panic_fmt::haf8695bfb6c52436
  │    6:        0x1089f0327 - rust_begin_unwind
  │    7:        0x108a1bc40 - core::panicking::panic_fmt::hfedd6515bcee8bb1
  │    8:        0x108984070 - core::result::unwrap_failed::he51c50f89cd6ddf5
  │    9:        0x10898792d - webrender_traits::api::RenderApi::set_root_pipeline::h4c242d85f833a2fc
  │   10:        0x106c9b580 - _$LT$compositing..compositor..IOCompositor$LT$Window$GT$$GT$::handle_browser_message::hfbc4a9dc7d4602f0
  │   11:        0x106c82a54 - _$LT$servo..Browser$LT$Window$GT$$GT$::handle_events::h15e4e3d1da942885
  │   12:        0x106ca9625 - servo::main::h6af425760c06b182
  │   13:        0x1089f172a - __rust_maybe_catch_panic
  │   14:        0x1089f0a34 - std::rt::lang_start::h6e7c6ad302fab11a
  │ ERROR:servo: called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) })
  │ stack backtrace:
  │    0:        0x1080f46ee - backtrace::backtrace::trace::h15d769766aa71ae0
  │    1:        0x1080f487c - backtrace::capture::Backtrace::new::hde1afc496271e12d
  │    2:        0x106caa666 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::he3b0f9713346a53e
  │    3:        0x1089f05f6 - std::panicking::rust_panic_with_hook::h8b9b25777425677b
  │    4:        0x1089f0454 - std::panicking::begin_panic::h69fe06a9d67b6165
  │    5:        0x1089f03c2 - std::panicking::begin_panic_fmt::haf8695bfb6c52436
  │    6:        0x1089f0327 - rust_begin_unwind
  │    7:        0x108a1bc40 - core::panicking::panic_fmt::hfedd6515bcee8bb1
  │    8:        0x108984070 - core::result::unwrap_failed::he51c50f89cd6ddf5
  │    9:        0x108987abb - webrender_traits::api::RenderApi::set_display_list::h1a91da94920c94ee
  │   10:        0x106e0a113 - layout_thread::LayoutThread::perform_post_style_recalc_layout_passes::h873d14182687ceb0
  │   11:        0x106e0501d - layout_thread::LayoutThread::handle_reflow::h3dc9096734abf992
  │   12:        0x106dfbcc3 - layout_thread::LayoutThread::handle_request_helper::h45a7767e5fa20613
  │   13:        0x106df4021 - _$LT$layout_thread..LayoutThread$u20$as$u20$layout_traits..LayoutThreadFactory$GT$::create::_$u7b$$u7b$closure$u7d$$u7d$::h994f01b5cfa1cc69
  │   14:        0x106d86407 - std::sys_common::backtrace::__rust_begin_short_backtrace::ha012796510649b84
  │   15:        0x106d87d97 - std::panicking::try::do_call::ha3734430072a78e7
  │   16:        0x1089f172a - __rust_maybe_catch_panic
  │   17:        0x106da6df4 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hf92b95d0408f806a
  │   18:        0x1089ed4c5 - std::sys::imp::thread::Thread::new::thread_start::h993b7e5cb7871f66
  │   19:     0x7fff95aed059 - _pthread_body
  │   20:     0x7fff95aecfd6 - _pthread_start
  └ ERROR:servo: called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) })

  ▶ CRASH [expected OK] /dom/nodes/Document-contentType/contentType/contenttype_txt.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ Can't allocate > MAX_VERTEX_TEXTURE_WIDTH per resource! (thread RenderBackend, at /Users/servo/.cargo/git/checkouts/webrender-c3596abe1cf4f320/caec344/webrender/src/gpu_cache.rs:240)
  │ 2017-06-08 06:23:42.451 servo[34667:1658180740] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x10559a6ee - backtrace::backtrace::trace::h15d769766aa71ae0
  │    1:        0x10559a87c - backtrace::capture::Backtrace::new::hde1afc496271e12d
  │    2:        0x104150666 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::he3b0f9713346a53e
  │    3:        0x105e965f6 - std::panicking::rust_panic_with_hook::h8b9b25777425677b
  │    4:        0x10563a634 - std::panicking::begin_panic::h0180222598f02c7d
  │    5:        0x1056bf92a - _$LT$webrender..gpu_cache..GpuDataRequest$LT$$u27$a$GT$$u20$as$u20$core..ops..Drop$GT$::drop::h39e0b81640f596f9
  │    6:        0x1056c2f20 - _$LT$webrender..prim_store..TextRunPrimitiveCpu$u20$as$u20$webrender..gpu_cache..ToGpuBlocks$GT$::write_gpu_blocks::hc5b7908553c4ba22
  │    7:        0x1056bcc87 - webrender::frame_builder::LayerRectCalculationAndCullingPass::handle_primitive_run::h3433b24e91763df0
  │    8:        0x1056b2abf - webrender::frame_builder::FrameBuilder::build::h82fe04f16c0fd444
  │    9:        0x1056d34b1 - webrender::render_backend::RenderBackend::render::hc7ba8677c3d5fe77
  │   10:        0x1056c8348 - webrender::render_backend::RenderBackend::run::h7d249d7284265167
  │   11:        0x105638c55 - std::sys_common::backtrace::__rust_begin_short_backtrace::h11923530aed18a36
  │   12:        0x10563b88d - std::panicking::try::do_call::h7dc0a5e886a59d78
  │   13:        0x105e9772a - __rust_maybe_catch_panic
  │   14:        0x1056597a6 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::h9765d4eb192e382f
  │   15:        0x105e934c5 - std::sys::imp::thread::Thread::new::thread_start::h993b7e5cb7871f66
  │   16:     0x7fff95aed059 - _pthread_body
  │   17:     0x7fff95aecfd6 - _pthread_start
  │ ERROR:servo: Can't allocate > MAX_VERTEX_TEXTURE_WIDTH per resource!
  │ called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) }) (thread main, at src/libcore/result.rs:859)
  │ called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) }) (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(2) }, at src/libcore/result.rs:859)
  │ stack backtrace:
  │    0:        0x10559a6ee - backtrace::backtrace::trace::h15d769766aa71ae0
  │    1:        0x10559a87c - backtrace::capture::Backtrace::new::hde1afc496271e12d
  │    2:        0x104150666 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::he3b0f9713346a53e
  │    3:        0x105e965f6 - std::panicking::rust_panic_with_hook::h8b9b25777425677b
  │    4:        0x105e96454 - std::panicking::begin_panic::h69fe06a9d67b6165
  │    5:        0x105e963c2 - std::panicking::begin_panic_fmt::haf8695bfb6c52436
  │    6:        0x105e96327 - rust_begin_unwind
  │    7:        0x105ec1c40 - core::panicking::panic_fmt::hfedd6515bcee8bb1
  │    8:        0x105e2a070 - core::result::unwrap_failed::he51c50f89cd6ddf5
  │    9:        0x105e2d92d - webrender_traits::api::RenderApi::set_root_pipeline::h4c242d85f833a2fc
  │   10:        0x104141580 - _$LT$compositing..compositor..IOCompositor$LT$Window$GT$$GT$::handle_browser_message::hfbc4a9dc7d4602f0
  │   11:        0x104128a54 - _$LT$servo..Browser$LT$Window$GT$$GT$::handle_events::h15e4e3d1da942885
  │   12:        0x10414f625 - servo::main::h6af425760c06b182
  │   13:        0x105e9772a - __rust_maybe_catch_panic
  │   14:        0x105e96a34 - std::rt::lang_start::h6e7c6ad302fab11a
  │ ERROR:servo: called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) })
  │ stack backtrace:
  │    0:        0x10559a6ee - backtrace::backtrace::trace::h15d769766aa71ae0
  │    1:        0x10559a87c - backtrace::capture::Backtrace::new::hde1afc496271e12d
  │    2:        0x104150666 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::he3b0f9713346a53e
  │    3:        0x105e965f6 - std::panicking::rust_panic_with_hook::h8b9b25777425677b
  │    4:        0x105e96454 - std::panicking::begin_panic::h69fe06a9d67b6165
  │    5:        0x105e963c2 - std::panicking::begin_panic_fmt::haf8695bfb6c52436
  │    6:        0x105e96327 - rust_begin_unwind
  │    7:        0x105ec1c40 - core::panicking::panic_fmt::hfedd6515bcee8bb1
  │    8:        0x105e2a070 - core::result::unwrap_failed::he51c50f89cd6ddf5
  │    9:        0x105e2dabb - webrender_traits::api::RenderApi::set_display_list::h1a91da94920c94ee
  │   10:        0x1042b0113 - layout_thread::LayoutThread::perform_post_style_recalc_layout_passes::h873d14182687ceb0
  │   11:        0x1042ab01d - layout_thread::LayoutThread::handle_reflow::h3dc9096734abf992
  │   12:        0x1042a1cc3 - layout_thread::LayoutThread::handle_request_helper::h45a7767e5fa20613
  │   13:        0x10429a021 - _$LT$layout_thread..LayoutThread$u20$as$u20$layout_traits..LayoutThreadFactory$GT$::create::_$u7b$$u7b$closure$u7d$$u7d$::h994f01b5cfa1cc69
  │   14:        0x10422c407 - std::sys_common::backtrace::__rust_begin_short_backtrace::ha012796510649b84
  │   15:        0x10422dd97 - std::panicking::try::do_call::ha3734430072a78e7
  │   16:        0x105e9772a - __rust_maybe_catch_panic
  │   17:        0x10424cdf4 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hf92b95d0408f806a
  │   18:        0x105e934c5 - std::sys::imp::thread::Thread::new::thread_start::h993b7e5cb7871f66
  │   19:     0x7fff95aed059 - _pthread_body
  │   20:     0x7fff95aecfd6 - _pthread_start
  └ ERROR:servo: called `Result::unwrap()` on an `Err` value: IoError(Error { repr: Custom(Custom { kind: NotFound, error: StringError("Bogus destination port.") }) })
```

### PR DESCRIPTION
**Pull request by pcwalton**
Closes #17230.

r? @jdm

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #___ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by MeFisto94** [comment]
If possible, I'd suggest querying the Vertex Texture Size dynamically and deduct the maximum glyph count from that, because technically recent hardware can do a much higher texture resolution natively and that way if webrender changes from hardcoded texture size limits to some hardware detection based mechanism, your code could adapt.
Technically for openGL (not sure if webrenderer uses Vulkan?) that would be as simple as calling `GLES10.glGetIntegerv(GL10.GL_MAX_TEXTURE_SIZE, maxTextureSize, 0);` 

But then I guess, unless webrender exposes this with a proper API, such a change doesn't make much sense and it'd be easy to adopt this code once / if ever webrender has dynamic texture sizes

File: components/layout/display_list/builder.rs [hunk 1]
```diff
@@ -124,6 +124,11 @@ static THREAD_TINT_COLORS: [ColorF; 8] = [
     },
 ];
 
+// An internal WebRender limit.
+//
+// See: https://github.com/servo/servo/issues/17230#issuecomment-564307277
+const MAX_GLYPHS_PER_TEXT_RUN: usize = 2000;
```
Comment (jdm) [inline components/layout/display_list/builder.rs]: Please file a WebRender issue about exposing an API for this.

**Review by jdm — APPROVED** [review]

**Comment by pcwalton** [comment]
I think the proper fix for this is probably on WebRender's side, honestly. I'll file an issue to that effect.

**Pull request merged**